### PR TITLE
Hessians!

### PR DESCRIPTION
**Pull request by andysim**
## Description
Adds analytic RHF Hessians.  Still much cleanup/fixing to be done, but I thought I'd open a PR to allow others to chime in.

## Todos
- [x] Add tests.
- [x] Hook into driver (with help from @loriab).
- [x] Add batching to allow fock derivatives to be computed without using too much memory.
- [x] Fix bug in spherical harmonic second derivative integrals.

## Questions
- [x]  Should we just merge this in when UHF is working, or should we also push to get the XC Hessian terms, @dgasmith / @robparrish ?  I don't know how involved the XC terms will be, but I'm happy to help get them cranked out.

## Status
- [x]  Ready to go

### CODE REVIEW

**Comment by andysim** [comment]
Btw, @dgasmith , I didn't `pull --rebase` this one because the sequential patching hits a merge conflict that I resolved some time ago (we'd both messed with scf_grad.cc at some point).  When this is ready to go I'll squash the commits.

**Comment by loriab** [comment]
How fitting, Fourth of July, and the Englishman brings Hessians state-side :-) These are very welcome, though. Thanks, @andysim. So narrowly lost the race with `1.0`.

**Comment by andysim** [comment]
I didn't think of that, @loriab.  These Hessians come in peace (although they could well impart the occasional seg fault).  I was within epsilon of having the code done last week (literally; I screwed up some orbital energy scaling term somehow) so I came very close to beating `1.0`.  Still, this code is almost useless until we get DF and KS terms in there, so no sense in hurrying it into a new release version.

**Comment by loriab** [comment]
Glad to hear it. :-) I'll be focusing on SciPy through the 17th so, besides the driver hook, let me know if you need input from me.

**Comment by dgasmith** [comment]
Eh, always a time for merging. Just trying to cut down on all the needless merging that goes on.

This looks great, I think a reference implementation for now is fine. We won't hook anything up, but it would good to have in the code. I think UHF/DFT/DF- are on a long enough time frame to not worry about in this particular PR. Especially since we move from something thats broken to something that works.

**Comment by andysim** [comment]
Ok, I agree with your idea @dgasmith.  I'll make sure the RHF stuff is working properly (I think there's still a small bug in the spherical harmonic hessians) and we can probably close this one out when all is working and more tests are added.  I know you're both busy, but if you get a chance could you take a look at `tests/scf-hess1` and see if you have any recommendations about implementing the call into the `frequencies` wrapper, please?

File: tests/scf-hess1/input.dat [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+set {
+  puream true
+  basis cc-pvdz
+}
+
+energy,wfn = energy('scf', return_wfn=True)
+psi4.scfhess(wfn)
+psi4_hess = np.asarray(wfn.hessian())
+
+if not np.allclose(psi3_hess, psi4_hess, atol=1E-7):
```
Comment (dgasmith) [inline tests/scf-hess1/input.dat]: You can use `compare_arrays` here which wraps `np.allclose`.

**Comment by dgasmith** [comment]
Add a new method in `share/python/procedures/proc_table.py:172` that links to a function in `share/python/procedures/proc_table.py:1710` (line number approximate) that should mimic `run_scf_gradient` right above that line.

Make sure to raise a Validation error for non-RHF and DFT references for now. The Hessian should then run assuming some untested code works as expected.

For frequencies: when we use finite difference Hessians its easy to assign different frequencies to different symmetries, without that it becomes much more difficult. You can guess which frequency applies to which frequency, but you cannot be certain in some cases.

@loriab Do you remember the conclusion of this? I think we might have to insert our new thermo module for this.

@bwb314 Hows the NumPy thermo module coming?

**Comment by loriab** [comment]
On the driver: Presently, `select_` methods (e.g., https://github.com/psi4/psi4/blob/master/share/python/procedures/proc.py#L127) are only used for post-scf, but it may be worthwhile to put such in for scf. Their purpose is to organize and filter availability based on reference (rhf, etc.), algorithm (conv, df, etc.), and module (irrelevant here- all scf). `gradient()` knows how to play nicely with `select_` methods, using analytic when avail. and falling back to findif if not. That's the behavior we want for hessian, too. Lines like https://github.com/psi4/psi4/blob/master/share/python/driver.py#L567 could also be absorbed into a `select_`. Basically, the procedures dict calls the `select_scf_grad/hess` and that passes along to `run_scf` which calls `scf_helper`.

On the vibrational analysis/symm sorting:
Andy and I were chatting about this after your talk, @dgasmith. He found the libmints matrix to transform the Cart Hessian back to symmetry blocked. So then can readily classify. That should do for the Abelian PGs. @bwb314 has been commissioned.

```
mints = MintsHelper(wfn.basisset())
cdsalcs = mints.cdsalcs(15, True, True)
that does it
U = cdsalcs.matrix()
well, the 15 should really be 0xF
Hsymm = U^T H U
```

**Comment by andysim** [comment]
Thanks for the helpful feedback.  I've implemented the driver code, but haven't checked it in yet.  I'm battling a small issue with non C1 calculations, but it should be easy to fix.  I've made the test a little prettier too, but wasn't aware of `compare_arrays` playing nicely with numpy arrays, so thanks for the info.  Also, the `0xF` that I mentioned in my chat with @loriab should be `0xFF` to get all irreps of the transformation matrix.

**Comment by andysim** [comment]
Ok, I think the RHF code is ready to go now.  It can run out of core, and is symmetry tolerant.  Adding UHF will take a quite a bit of time to debug.  Density fitting should be quite fast to code up, as it's only really needed in the second derivatives and the G terms in the response code.  Per the above discussion, we can save those tasks for another day.  I will let these tests run, then you can merge whenever is most convenient for the new CMake infrastructure.  Also, let me know if you'd rather me squash all these commits, which go back all the way to November 2015 and could cause some issues.

**Comment by dgasmith** [comment]
Sounds good. I wouldn't bother squashing these they look fairly distinct. Git now allows us to squash upon merge these days if needed.

**Comment by andysim** [comment]
Travis identified a problem.  I'd tested the Hessian itself, but hadn't ensured that the code plays nicely with `frequency()`.  Turns out it doesn't so I'll see if I can come up with a quick fix.  If not, we'll have to wait for the py-side frequency analysis before pulling this in.

**Comment by loriab** [comment]
Looks like you tried to fix the test cases that formerly tested scf findif-by-grad freq and now are accessing the analytic freq by adding `der_type=1`. Sounds like a good plan to me, as we don't want this PR lingering and getting out of sync. The `der_type=1` didn't work. Perhaps try `dertype=1`, which I think is what the driver acts on.

**Comment by andysim** [comment]
By hooking the hessian call up to the analytic RHF code, I unleashed the error message in the frequency tests that use DF and, as you said, used the incorrect keyword to request gradients.  I will fix those tests when I get a chance (probably the flight back on Monday - I'm at a conference now).  On my flight here I wrote some cheesy frequency code to plug the gap in our capabilities, which may or may not be useful::

    import numpy as np
    from math import pi
    mol = ref_wfn.molecule()
    natoms = mol.natom()
    masses = np.zeros(natoms)
    for atom in range(natoms):
        masses[atom] = mol.mass(atom)
    m = np.diagflat( np.repeat( np.divide(1.0, np.sqrt(masses)),3))
    mwhess = np.dot(m, np.dot(H, m))
    fcscale = psi_hartree2J/(psi_bohr2m * psi_bohr2m * psi_amu2kg);
    fc = fcscale*np.linalg.eigvalsh(mwhess)
    sqrtfc = np.sqrt(np.abs(fc))
    scalefac = 1.0/(2.0 * pi * psi_c * 100.0)
    freqs = scalefac*np.where(fc > 0.0, sqrtfc, -sqrtfc)
    freqvec = psi4.Vector(3*natoms-6)
    n = 0
    for f in freqs[6:]:
        freqvec.set(n, f)
        n += 1
    ref_wfn.set_frequencies(freqvec)

**Comment by andysim** [comment]
I should add, the `der_type` keywords were already in the tests.  I'm not sure how pervasive they are, but we should probably weed them all out.

**Comment by dgasmith** [comment]
@andysim Updated your algorithm, please review.

**Comment by andysim** [comment]
Thanks, @dgasmith.  I just pushed a fix to force finite differences in the test cases that need DF SCF frequencies.  Will revert those changes when we get analytic Hessians for that method.  Hopefully this will fix all broken tests; the code is ready to go if it works.

**Comment by andysim** [comment]
I had a long day of travel back from the conference, and used the ensuing boredom as an excuse to code up the DF second derivatives. The code is not completely terrible, and gives correct answers. We need an efficient out of core code eventually, but what do you think about adding it to this PR?

**Comment by dgasmith** [comment]
Sure, more templates to build off of in the future are always welcome. FYI: Lori and I are at a conference so probably not much movement on anything Psi4 related until next week.

**Comment by andysim** [comment]
OK, the DF Hessians (including proper orbital response terms, with the SCF fitting basis) are in there now.  There's a problem with the `OUT_OF_CORE` algorithm somewhere in libJK, but `PK`/`DIRECT` work just fine so I'm not sure it's worth worrying about too much.  I saw that `g2` uses `OUT_OF_CORE`; is there a reason for this, @loriab?  If it's OK to switch to `PK` for that method, I think that's a reasonable fix for now.

**Comment by loriab** [comment]
I think pk for g2 would be fine. Anything but df.  @edeprince3 can comment if not.

**Comment by edeprince3** [comment]
I’m certain I used PK when I first wrote that G2 routine.  I don’t see why it can’t be used now.

File: src/lib/libmints/eribase.cc [hunk 2]
```diff
@@ -1808,21 +1808,21 @@ TwoElectronInt::TwoElectronInt(const IntegralFactory* integral, int deriv, bool
...
                         "     In a fresh object directory, reconfigure libderiv for higher angular momentum, then recompile.\n");
         throw LimitExceeded<int>("ERI - libderiv cannot handle angular momentum this high.\n"
                                  "In a fresh object directory, reconfigure libderiv for higher angular momentum, then recompile.",
                                  LIBDERIV_MAX_AM1, max_am, __FILE__, __LINE__);
     }
-    else if (deriv_ == 2 && max_am >= LIBDERIV_MAX_AM12) {
+    else if (deriv_ == 2 && max_am > LIBDERIV_MAX_AM12) {
         outfile->Printf( "ERROR: ERI - libderiv cannot handle angular momentum this high.\n"
                         "       In a fresh object directory, reconfigure libderiv for higher angular momentum, then recompile.\n");
         throw LimitExceeded<int>("ERI - libderiv cannot handle angular momentum this high.\n"
```
Comment (andysim) [inline src/lib/libmints/eribase.cc]: Does this look correct to you, @jturney?  The tests all seem to pass with the new checks in place, but I wanted to run it by you to make sure the equality condition wasn't deliberate.

Comment (jturney) [inline src/lib/libmints/eribase.cc]: Yeah, this looks right.  If the tests pass then it's good.

Comment (dgasmith) [inline src/lib/libmints/eribase.cc]: This would explain some curiosities that I have been seeing when hitting the AM limit. Great that we can fix this.

**Comment by dgasmith** [comment]
@andysim Whats going wrong with out of core?

File: src/lib/libmints/eribase.cc (hunk 2 above)
Comment (andysim) [inline src/lib/libmints/eribase.cc]: It gives us a decision to make; to leave the AM at 4 on Travis and allow higher angular momentum in the tests, or to reduce the build AM to 3.

**Comment by andysim** [comment]
The code is correct, but I haven't put much effort into making it memory efficient.  The [second derivative terms](https://github.com/andysim/psi4/blob/hessians/src/bin/scfgrad/jk_grad.cc#L1593-L2336) allocate some large matrices up top (could still do some pretty large calculations with these in core), and the [Fock derivative terms](https://github.com/andysim/psi4/blob/hessians/src/bin/scfgrad/response.cc#L443-L779) also allocate large chunks.  The Fock derivative code does at least handle the 3*N*nso*nso results in a batched way.  The biggest problem is that I lazily recompute some quantities that @robparrish wrote efficient routines for handling out of core, particularly `(A|ij)` and `(A|B)^{-1}(B|ij)`.  If anybody wants to have a stab a refactoring the current code, be my guest.  I'll be able to get to it some day, but not in the next couple of months.  The aim of this implementation is really just to have something that works, to serve as a guideline for our efforts to get Hessians for many other methods.  I know that @bozkaya and @lothian were interested in getting Hessians into their codes, and I'm more than happy to work with them to get this code adapted for their needs.

File: src/lib/libmints/eribase.cc (hunk 2 above)
Comment (dgasmith) [inline src/lib/libmints/eribase.cc]: I vote for leaving it at 4, we have to gimp a couple test cases that could really use higher AM (such as extrapolation testing). I don't think the difference between AM 3 and 4 is much in terms of compilation cost.

Comment (andysim) [inline src/lib/libmints/eribase.cc]: I completely agree.  The scf-freq1 test is reasonable with AM=4, but is hard to really test for AM=3.

**Comment by dgasmith** [comment]
Ah, ok it just sounded like there was a bug in the JK `out_of_core` algorithm. I wouldn't worry about that too much, we have a new DF class coming along that will be able to handle these kinds of things very efficiently.

**Comment by andysim** [comment]
Ah, yes, that out of core problem.  It's a strange one that manifests itself in a seg fault for out of core calls to JK but, apparently, only when the L and R matrices are not the same (as is the case in CPHF).  I'll just call PK instead, and add an error if out of core flags are requested.  I'll squash this last flurry of commits and then this one is good to go.

**Comment by dgasmith** [comment]
Thats interesting, I actually used `out_of_core` for non-symmetric densities up until recently as PK could not handle these. Only recent changes by Jerome to the PK code allow this to work. 

Looks like OOC still works for SOSCF, do you have time to investigate the error a little bit? Shouldn’t be happening.

Cheers,
Daniel Smith

**Comment by andysim** [comment]
@dgasmith. I think this is a PEBKAC error.  I was passing in desymmetrized matrices, which is fine for PK and DF, which both use C1 symmetry under the hood.  The fix for `OUT_OF_CORE` is to add symmetry back into the L and R matrices, before calling `jk->compute()`.  There's no problem with the JK code, and I'll see if I can add a simple fix for out of core.

**Comment by andysim** [comment]
Hold off on this one, and I'll try to resolve the symmetry issue with `OUT_OF_CORE`.

**Comment by dgasmith** [comment]
@andysim Yea, everything else was ready to go and we agreed internally that with the CMake rewrite it would be best off if we rebased them ourselves. Getting *very* hard to do.

**Comment by andysim** [comment]
OK, another change of plan.  The only reason to symmetrize the response terms is to be able to use JK instances that use symmetry under the hood (which, AFAIK, is only `OUT_OF_CORE`).  I'm half way through doing this, but it makes the code very messy and I only just realized that `OUT_OF_CORE` can't even handle non-symmetric density matrices.  I think the best plan at this point is for me to abandon the symmetrization, clean the code up, squash the last few commits and add a py-side error message if `OUT_OF_CORE` is requested.  Does anybody disagree.  If more important integral technologies were affected, I'd say we go ahead an symmetrize.  As it is, we'd be introducing a lot of messy code, just to cover an unimportant corner case.

**Comment by mac-psinet** [comment]
@andysim `OUT_OF_CORE` can use symmetry under the hood. We can look at symmetrizing it later, but shouldn't it be one line of code using the matrix `transform` function? The JK builder will usually just transform is back, but still worthwhile for compatibility IMO.

**Comment by andysim** [comment]
I agree that compatibility is nice, and will press ahead with adding the symmetry back in if that's the consensus.  My hesitation to do so stems from the fact that only `OUT_OF_CORE` can use the symmetry, and it's the worst algorithm we have in JK.  Symmetrizing is a lot more work than using `transform()` to transform the matrices (which doesn't seem to play nicely with mixed symmetric/c1 matrices), because the perturbations themselves must also be transformed to SALCs.  None of that is hard, but it does make the response code messier; I currently check to see if the JK engine runs in C1 anyway, and avoid symmetrization for those cases.

**Comment by mac-psinet** [comment]
Sure, we can always fix it later. Just looked like it would be an easy fix.

**Comment by dgasmith** [comment]
You know the JK object owns a AOTOSO object. I wonder if should elevate the JK object so that it can take either an AO or SO representation, transform it as needed, and return the input. A relatively small amount of logic all things considered.

Anyhow, is this ready to go Andy?

**Comment by andysim** [comment]
I'm using the AOTOSO object for the transformations.  I'll clean up what I have and push it later today.  I need to do a little squashing and removal of some junk before this goes in.

**Comment by andysim** [comment]
Ok, this one should be ready to go, if all tests pass.  I squashed a few of the commits, so if you forked this code at all be careful pushing upstream.  I ended up putting the symmetry that I have so far into the code, but it's not used.  If we want symmetry to work, here's what needs to happen.

-  [Activate the symmetry code](https://github.com/andysim/psi4/blob/hessians/src/bin/scfgrad/response.cc#L1090) and run it in C1 symmetry, which should all be good.

- Then for something simple like water in Cs symmetry you can [verify that the B vectors are good](https://github.com/andysim/psi4/blob/hessians/src/bin/scfgrad/response.cc#L1411) (remember that they'll be ordered differently to the C1 case because of the SALC ordering being different) but the CPHF doesn't converge, so we need to figure out the problem there.

- After that, there's just [another small cut and paste exercise](https://github.com/andysim/psi4/blob/hessians/src/bin/scfgrad/response.cc#L1510) to get all terms symmetrized, albeit in a slightly janky way.

- Once we figure out why CPHF isn't working, we need to make `OUT_OF_CORE` work with non-symmetric densities and [remove the check](https://github.com/andysim/psi4/blob/hessians/share/python/procedures/proc.py#L1728) in the driver that stops these integrals from being allowed.

If we end up adding a new JK object that uses symmetry, the above steps will be worth it.  Just to get `OUT_OF_CORE` working, not so much unless anybody already knows why it doesn't work and can easily fix it.  This is why I've been back and forth about including the symmetry code; I eventually decided to leave it in as a reasonable starting point for future efforts.

**Comment by dgasmith** [comment]
To be clear the JK OOC algorithm *does* work with non-symmetric density cases. See `/tests/soscf2` for an example.

Everything looks good and ill pull it in when Travis goes through. Ignore the Distelli stuff for now.

**Comment by andysim** [comment]
Weird.  I'm not sure what was going on when I tried it before, but the K matrices for non symmetric perturbations were just wrong.  Switching to a different algorithm fixed it.

**Comment by andysim** [comment]
I'd broken opt-irc-2 accidentally (not part of quicktests) so I just added some Py-side code to write out H to an ASCII file, which fixes it.

**Comment by dgasmith** [comment]
This passed long test on box! Thanks Andy for all your hard work.

If this is ready ill go ahead and pull this in.

**Comment by andysim** [comment]
Yup, ready to go.  Glad to hear that nothing else got screwed up.  There's a lot of room for improvement in this implementation, but it should work for some reasonably large systems.

**Comment by loriab** [comment]
@andysim, I noticed that [`fd-freq-energy-large`](https://github.com/psi4/psi4/blob/master/tests/fd-freq-energy-large/input.dat) was timing out locally. It has various tests; the first is set up for findif-by-E, the rest are unspecified but partial freqs. The test was doing the first by findif fine, then going into `SCF HESS` and timing out.

The intent of the test is clear- to test findif-by-E, so I've made that explicit. My real question for you is: is that testing change contrary to anything you intended? It doesn't look like you touched the test in your PR. (And yes, looks like some re-routing from analytic to findif is in order if a partial freq is requested.)

**Comment by andysim** [comment]
Nope, just an oversight on my part.  Good catch.  Most of the changes to the tests that I made were a result of high precision checks failing (3 point f.d. vs. analytic can be a little noisy), so I was just doing trial and error to figure out where things needed to be updated.  I must have missed this one, because I just saw that it timed out on my machine.  Sorry about that.

**Comment by loriab** [comment]
Great, thanks @andysim, I'll fix it up.

**Pull request merged**